src/pip/_internal/vcs/git.py

```diff
@@ -104,17 +104,19 @@ def check_rev_options(self, dest, rev_options):
             )
             return rev_options
 
-    def check_version(self, dest, rev_options):
+    def is_commit_id_equal(self, dest, name):
         """
-        Compare the current sha to the ref. ref may be a branch or tag name,
-        but current rev will always point to a sha. This means that a branch
-        or tag will never compare as True. So this ultimately only matches
-        against exact shas.
+        Return whether the current commit hash equals the given name.
 
         Args:
-          rev_options: a RevOptions object.
+          dest: the repository directory.
+          name: a string name.
         """
-        return self.get_revision(dest).startswith(rev_options.arg_rev)
+        if not name:
+            # Then avoid an unnecessary subprocess call.
+            return False
+
+        return self.get_revision(dest) == name
```

Comment:
I don't like the fact that `2b466712a067a171e8d5a0fe6210e004cfe103a8` won't be equal to `2b466712a`.

Would self.get_revision(dest) == self.run_command(['rev-parse',name], show_stdout=False, cwd=location` work ?

Reply:
After thinking some more I guess the point of this PR is that `2b466712a` could also be the name of a branch so my remark is moot.
If one wants to avoid an useless update, one should specify the whole commit hash. A additionnal note in https://pip.pypa.io/en/stable/reference/pip_install/#git might be welcome :)

Reply:
But the thing you're checking is user-supplied, and requiring people to provide a full hash is pretty user-unfriendly. We should match git's behaviour (which is to allow prefixes as long as they don't match a branch or tag, as I understand it).

Disclaimer: I've never used this feature, and don't anticipate a need to. So if people who actually use it are happy with requiring a full hash, then their opinions should override mine.

Reply:
Right. For example, if you look at the original docstring, you'll see that the function isn't ever supposed to return True for branches or tags. A more realistic example might be an integer tag like "10". An accidental SHA prefix match could cause the tag not to be installed.

Specifying a SHA prefix should still work to install a SHA. It just won't incorrectly "shadow" a branch or tag, which is what this PR addresses.

Reply:
@pfmoore Providing a partial SHA still works. The user isn't required to provide a full hash. This PR just addresses an optimization code path. The optimization was implemented not quite correctly. It was implemented so that if a user specified a tag that accidentally matches the beginning of a SHA, the tag wouldn't get installed (because the optimization bypasses fetching new tags).

Reply:
@pfmoore specifying a short hash still work but pip will update the source in this case instead of assuming equality.
Cf https://gist.github.com/xavfernandez/47ac526e0c6bd0ada872deba2c3c2c05 for the difference between pip 9.0.1 and this PR.

Reply:
Thanks for taking the time to do those examples, @xavfernandez.

Reply:
No problem, if @pfmoore agrees I'd say the only thing missing before merging would be a small note in https://github.com/pypa/pip/blob/master/docs/reference/pip_install.rst#git if it's not too much to ask :)

Reply:
Sure, would be happy to do that. I'll see if I can do that now.

Reply:
Thanks for clarifying, yes I'm happy for this to go in with the note as requested.